File: DemoServer/AdditionalLanguages/python/textSearch/highlightQueryResultsCustomized/highlightQueryResultsCustomized.py
from ravendb.documents.queries.highlighting import Highlightings, HighlightingOptions
from demo_example import RunParamsBase, Example
from models import Employee

#region Usings
from typing import Optional
from ravendb import AbstractIndexCreationTask
from ravendb.documents.indexes.definitions import (
    FieldStorage,
    FieldIndexing,
    FieldTermVector,
)
#endregion

class RunParams(RunParamsBase):
    def __init__(
        self,
        fragment_length: int = None,
        fragment_count: int = None,
        tag1: str = None,
        tag2: str = None,
        tag3: str = None,
        tag4: str = None,
    ):
        self.fragment_length = fragment_length
        self.fragment_count = fragment_count
        self.tag1 = tag1
        self.tag2 = tag2
        self.tag3 = tag3
        self.tag4 = tag4

class DataToShow:
    def __init__(self, document_id: str = None, index_field: str = None, fragment: str = None):
        self.document_id = document_id
        self.index_field = index_field
        self.fragment = fragment

#region Demo
#region Step_1
class EmployeesDetails(AbstractIndexCreationTask):
#endregion
    
    def __init__(self):
        #region Step_2
        super().__init__()
        self.map = (
            "docs.Employees.Select(employee => new {"
            "    Title = employee.Title,"
            "    Notes = employee.Notes[0]"
            # employee.Notes is a string array,
            # indexing only the first element for this example
            "})"
        )
        #endregion
        
        #region Step_3
        self._store("Title", FieldStorage.YES)
        self._index("Title", FieldIndexing.SEARCH)
        self._term_vector("Title", FieldTermVector.WITH_POSITIONS_AND_OFFSETS)

        self._store("Notes", FieldStorage.YES)
        self._index("Notes", FieldIndexing.SEARCH)
        self._term_vector("Notes", FieldTermVector.WITH_POSITIONS_AND_OFFSETS)
        #endregion
#endregion

class HighlightQueryResultsCustomized(Example):
    def run(self, run_params: RunParams):
        fragmentLength = run_params.fragment_length or 100
        fragmentCount = run_params.fragment_count or 1

        tag1 = run_params.tag1 or "+++"
        tag2 = run_params.tag2 or "+++"
        tag3 = run_params.tag3 or "<<<"
        tag4 = run_params.tag4 or ">>>"
        EmployeesDetails().execute(self.document_store_holder.store())

        #region Demo
        title_highlightings: Optional[Highlightings] = None

        def __title_highlight_callback(result):
            nonlocal title_highlightings
            title_highlightings = result
        
        notes_highlightings: Optional[Highlightings] = None

        def __notes_highlight_callback(result):
            nonlocal notes_highlightings
            notes_highlightings = result
        
        with self.document_store_holder.store().open_session() as session:
            #region Step_4
            tags_to_use1 = HighlightingOptions()
            tags_to_use1.pre_tags = [tag1]
            tags_to_use1.post_tags = [tag2]

            tags_to_use2 = HighlightingOptions()
            tags_to_use2.pre_tags = [tag3]
            tags_to_use2.post_tags = [tag4]
            #endregion

            #region Step_5
            employees_results = list(
                session.query_index_type(EmployeesDetails, Employee)
                .highlight(
                    "Title",
                    fragmentLength,
                    fragmentCount,
                    __title_highlight_callback,
                    tags_to_use1,
                )
                .highlight(
                    "Notes",
                    fragmentLength,
                    fragmentCount,
                    __notes_highlight_callback,
                    tags_to_use2,
                )
                .search("Title", "manager")
                .search("Notes", "sales")
            )
            #endregion

            #region Step_6
            if len(employees_results) > 0:
                employee_id = employees_results[0].Id
                title_fragments = title_highlightings.get_fragments(employee_id)
                notes_fragments = notes_highlightings.get_fragments(employee_id)
            #endregion
        #endregion
        
        highlight_results = []
        for employee in employees_results:
            title_fragments = title_highlightings.get_fragments(employee.Id)

            for item in title_fragments:
                item_results = DataToShow(employee.Id, title_highlightings.field_name, item)
                highlight_results.append(item_results)

            notes_fragments = notes_highlightings.get_fragments(employee.Id)
            for item in notes_fragments:
                item_results = DataToShow(employee.Id, notes_highlightings.field_name, item)
                highlight_results.append(item_results)

        return sorted(highlight_results, key=lambda x: x.index_field, reverse=True)

File: DemoServer/AdditionalLanguages/php/textSearch/highlightQueryResultsCustomized/HighlightQueryResultsCustomized.php
<?php

namespace RavenDB\Demo\textSearch\highlightQueryResultsCustomized;

//region Usings
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Indexes\FieldIndexing;
use RavenDB\Documents\Indexes\FieldStorage;
use RavenDB\Documents\Indexes\FieldTermVector;
use RavenDB\Documents\Queries\Highlighting\HighlightingOptions;
use RavenDB\Documents\Queries\Highlighting\Highlightings;
//endregion

//region Demo
//region Step_1
class EmployeesDetails extends AbstractIndexCreationTask
{
//endregion

    public function __construct()
    {
        parent::__construct();

        //region Step_2
        $this->map = "docs.Employees.Select(employee => new {" .
            "    Title = employee.Title," .
            "    Notes = employee.Notes[0]" .
            // employee.Notes is a string array,
            // indexing only the first element for this example
            "})";
        //endregion

        //region Step_3
        $this->store("Title", FieldStorage::yes());
        $this->index("Title", FieldIndexing::search());
        $this->termVector("Title", FieldTermVector::withPositionsAndOffsets());

        $this->store("Notes", FieldStorage::yes());
        $this->index("Notes", FieldIndexing::search());
        $this->termVector("Notes", FieldTermVector::withPositionsAndOffsets());
        //endregion
    }
}
//endregion

class HighlightQueryResultsCustomized
{
    public function __invoke(RunParams $runParams)
    {
        $fragmentLength = $runParams->getFragmentLength() ?? 100;
        $fragmentCount = $runParams->getFragmentCount() ?? 1;

        $tag1 = $runParams->getTag1() ?? "+++";
        $tag2 = $runParams->getTag2() ?? "+++";
        $tag3 = $runParams->getTag3() ?? "<<<";
        $tag4 = $runParams->getTag4() ?? ">>>";

        $titleHighlightings = new Highlightings();
        $notesHighlightings = new Highlightings();
        //region Demo
        $employeesResults = [];

        $session = DocumentStoreHolder::getStore()->openSession();
        try {

        //region Step_4
        $tagsToUse1 = new HighlightingOptions();
            $tagsToUse1->setPreTags([ $tag1 ]);
            $tagsToUse1->setPostTags([ $tag2 ]);

            $tagsToUse2 = new HighlightingOptions();
            $tagsToUse2->setPreTags([ $tag3 ]);
            $tagsToUse2->setPostTags([ $tag4 ]);
            //endregion

            //region Step_5
            $employeesResults = $session->query(Employee::class, EmployeesDetails::class)
                ->highlight("Title", $fragmentLength, $fragmentCount, $tagsToUse1, $titleHighlightings)
                ->highlight("Notes", $fragmentLength, $fragmentCount, $tagsToUse2, $notesHighlightings)
                ->search("Title", "manager")
                ->search("Notes", "sales")
                ->toList();
            //endregion

            //region Step_6
            if (count($employeesResults) > 0) {
                $employeeId = $employeesResults[0]->getId();
                $titleFragments = $titleHighlightings->getFragments($employeeId);
                $notesFragments = $notesHighlightings->getFragments($employeeId);
            }
            //endregion
        } finally {
            $session->close();
        }
        //endregion

        $highlightResults = [];

        /** @var Employee $employee */
        foreach ($employeesResults as $employee) {
            $titleFragments = $titleHighlightings->getFragments($employee->getId());

            foreach ($titleFragments as $item) {
                $itemResults = new DataToShow();
                $itemResults->setDocumentId($employee->getId());
                $itemResults->setIndexField($titleHighlightings->getFieldName());
                $itemResults->setFragment($item);

                $highlightResults[] = $itemResults;
            }

            $notesFragments = $notesHighlightings->getFragments($employee->getId());
            foreach ($notesFragments as $item) {
                $itemResults = new DataToShow();
                $itemResults->setDocumentId($employee->getId());
                $itemResults->setIndexField($notesHighlightings->getFieldName());
                $itemResults->setFragment($item);

                $highlightResults[] = $itemResults;
            }
        }

        usort($highlightResults, function(DataToShow $a, DataToShow $b) {
            return $a->getIndexField() < $b->getIndexField();
        });
        return $highlightResults;
    }
}

class DataToShow
{
    private ?string $documentId = null;
    private ?string $indexField = null;
    private ?string $fragment = null;

    public function getDocumentId(): ?string
    {
        return $this->documentId;
    }

    public function setDocumentId(?string $documentId): void
    {
        $this->documentId = $documentId;
    }

    public function getIndexField(): ?string
    {
        return $this->indexField;
    }

    public function setIndexField(?string $indexField): void
    {
        $this->indexField = $indexField;
    }

    public function getFragment(): ?string
    {
        return $this->fragment;
    }

    public function setFragment(?string $fragment): void
    {
        $this->fragment = $fragment;
    }
}
